ethers-providers/src/rpc/transports/ws/manager.rs

```diff
@@ -0,0 +1,400 @@
+        }
+    }
+
+    fn req_fail(&mut self, id: u64, error: JsonRpcError) {
+        // pending fut is missing, this is fine
+        if let Some(req) = self.reqs.remove(&id) {
+            // pending fut has been dropped, this is fine
+            let _ = req.channel.send(Err(error));
+        }
+    }
+
+    fn handle(&mut self, item: PubSubItem) {
+        match item {
+            PubSubItem::Success { id, result } => self.req_success(id, result),
+            PubSubItem::Error { id, error } => self.req_fail(id, error),
+            PubSubItem::Notification { params } => self.subs.handle_notification(params),
+        }
+    }
+
+    #[tracing::instrument(skip(self, params, sender))]
+    fn service_request(
+        &mut self,
+        id: u64,
+        method: String,
+        params: Box<RawValue>,
+        sender: oneshot::Sender<Response>,
+    ) -> Result<(), WsClientError> {
+        let in_flight = InFlight { method, params, channel: sender };
+        let req = in_flight.serialize_raw(id)?;
+
+        // Ordering matters here. We want this block above the unbounded send,
+        // and after the serialization
+        if in_flight.method == "eth_subscribe" {
+            self.subs.service_subscription_request(id, in_flight.params.clone())?;
+        }
+
+        // Must come after self.subs.service_subscription_request. Do not re-order
+        tracing::debug!("Dispatching request to backend");
+        self.backend.dispatcher.unbounded_send(req).map_err(|_| WsClientError::DeadChannel)?;
+
+        self.reqs.insert(id, in_flight);
+        Ok(())
+    }
+
+    fn service_instruction(&mut self, instruction: Instruction) -> Result<(), WsClientError> {
+        match instruction {
+            Instruction::Request { method, params, sender } => {
+                let id = self.next_id();
+                self.service_request(id, method, params, sender)?;
+            }
+            Instruction::Unsubscribe { id } => {
+                if let Some(req) = self.subs.end_subscription(id.low_u64()) {
+                    self.backend
+                        .dispatcher
+                        .unbounded_send(req)
+                        .map_err(|_| WsClientError::DeadChannel)?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn spawn(mut self) {
+        let fut = async move {
+            let result = loop {
+                // We bias the loop so that we always handle messages before
+                // reconnecting, and always reconnect before
+                select_biased! {
+                    item_opt = self.backend.to_handle.next() => {
+                        match item_opt {
+                            Some(item) => self.handle(item),
+                            // Backend is gone, so reconnect
+                            None => if let Err(e) = self.reconnect().await {
+                                break Err(e);
+                            }
```

Comment:
I could totally be wrong here because it wasn't obvious to which component `to_handle` is wired here. but it looks like the other half belongs to `WsBackend`?

I believe there's an issue with how the `WsBackend` task shuts down, which could cause two reconnects if the task exits because of an error?
first the channel will be None, then the  `self.backend.error` receives the error message and reconnects again?

Reply:
> I believe there's an issue with how the WsBackend task resolves, which could cause two reconnects if the task exits because of an error?

No, only 1 select branch is reachable in any pass through the loop. So if the `to_handle` channel is dead, the WS will reconnect. Then by the time the loop restarts, the `BackendDriver` will be different, and the error channel will be fresh and good

Reply:
> I could totally be wrong here because it wasn't obvious to which component to_handle is wired here. but it looks like the other half belongs to WsBackend?

the naming convention is you send a chunk of work to the handler via `handler`. the handler accepts work that it should handle via `to_handle`

same convention with `dispatcher` and `to_dispatch`. you send to the dispatching task via a `dispatcher` channel. the dispatcher accepts the requests via `to_dispatch`

Reply:
you can think of the `BackendDriver` as control or ownership of the running task. It has the ability to shut down that task, send/receive messages from it, and dropping the driver causes the task to shut down automatically

The communication flow is

client ----instructions---> manager
manager -----requests-----> backend
backend <-----WS Protocol-----> server
backend -----PubSubItem----> manager
manager ---sub notifications---> sub channels
manager ---sub channels---> client
manager ---responses----> request future channels

this would go well in a swim lanes but I don't have time to make one rn